src/arizona_template_compiler.erl

```diff
@@ -0,0 +1,646 @@
+        ({K, {text, _KLoc, KTxt}}) ->
+            {binary_to_existing_atom(K, utf8),
+              merl:quote(iolist_to_binary(["<<\"", KTxt, "\"/utf8>>"]))}
+    end, State#state.attributes),
+    AttrsTree = merl:tsubst(ExprTree, AttrsEnv),
+    Macros = State#state.macros,
+    MacrosEnv = [{K, merl:term(V)} || K := V <- Macros],
+    MacrosTree = merl:tsubst(AttrsTree, MacrosEnv),
+    AllVars = merl:template_vars(merl:template(MacrosTree)),
+    case AllVars -- maps:keys(Macros) of
+        [] when Eval ->
+            try
+                Form = erl_syntax:revert(merl:tree(MacrosTree)),
+                {value, Value, _} = erl_eval:exprs([Form], []),
+                {text, arizona_html:to_safe(Value)}
+            catch
+                _:_ ->
+                    throw({invalid_macro_value, Loc, State})
+            end;
+        Vars ->
+            VarsSubst = [{Var, var_form(Var)} || Var <- Vars],
+            Env = [{subst, merl:subst(MacrosTree, VarsSubst)}],
+            [Form] = erl_syntax:revert_forms([merl:qquote(~"_@subst", Env)]),
+            NormExpr = iolist_to_binary(erl_pp:expr(Form)),
+            Forms = [merl:quote(<<"fun(Assigns) -> ", NormExpr/binary, " end">>)],
+            {value, Fun, _} = erl_eval:exprs(Forms, []),
+            {expr, expr(Fun, Vars, State)}
+    end.
+
+var_form(Var) ->
+    VarName = atom_to_binary(Var, utf8),
+    merl:quote(<<"maps:get(", VarName/binary, ", Assigns)">>).
+
+expr(Fun, Vars, State) ->
+    #{
+        id => changeable_id(State),
+        function => Fun,
+        vars => Vars
+     }.
+
+compile_block(#{name := Name} = Block, Loc, T, State) ->
+    case block_mod_fun(Name, State) of
+        {ok, {Mod, Fun}} ->
+            Attrs = maps:get(attributes, Block),
+            Elems = expand_block(Mod, Fun, Attrs, State),
+            Elems ++ compile_elems(T, incr_index(changeable_count(Elems, 0), State));
+        {error, Reason} ->
+            throw({Reason, Loc, State})
+    end.
+
+block_mod_fun(Name, State) ->
+    try
+        {Mod, Fun} = case binary:split(Name, <<":">>) of
+            [M, F] ->
+                {binary_to_existing_atom(M, utf8), binary_to_existing_atom(F, utf8)};
+            [F] ->
+                {State#state.module, binary_to_existing_atom(F, utf8)}
+        end,
+        case erlang:function_exported(Mod, Fun, 1) of
+            true ->
+                {ok, {Mod, Fun}};
+            false ->
+                {error, undef_block_fun}
+        end
+    catch
+        _:_ ->
+            {error, invalid_block_name}
+    end.
+
+expand_block(Mod, Fun, Attrs, State) ->
+    case erlang:apply(Mod, Fun, [State#state.macros]) of
+        {ok, {[{tag, Loc, #{is_stateful := true} = Tag}], AttrsMacros}} ->
+            AllNormAssigns = norm_block_assigns(Attrs, AttrsMacros, State),
+            Macros = #{K => Txt || K := {text, Txt} <- AllNormAssigns},
+            NormAssigns = #{K => {expr, Expr} || K := {expr, Expr} <- AllNormAssigns},
```

Comment:
That's not true. `{text, Txt}` elements are valid macros only if the `K` `is_map_key` of the `AttrsMacros`. So, `NormAssigns` can contain {text, Txt} elements. I'll look at this.

Reply:
> [!NOTE]
>`K` is the key in the block, then it should be normalized to macros keys.

Reply:
After revisiting it, I think it is ok. If no variable is provided, there is no problem with considering it a macro because it will never change. I'll be entirely sure of this when creating the new renderer module.